core/src/no/group15/playmagic/ecs/systems/CollisionSystem.kt

```diff
@@ -0,0 +1,171 @@
+package no.group15.playmagic.ecs.systems
+
+import com.badlogic.ashley.core.Engine
+import com.badlogic.ashley.core.Entity
+import com.badlogic.ashley.core.EntitySystem
+import com.badlogic.ashley.core.Family
+import com.badlogic.ashley.signals.Listener
+import com.badlogic.ashley.signals.Signal
+import com.badlogic.ashley.utils.ImmutableArray
+import com.badlogic.gdx.math.*
+import ktx.ashley.allOf
+import ktx.ashley.get
+import ktx.ashley.mapperFor
+import no.group15.playmagic.ecs.components.ColliderShape
+import no.group15.playmagic.ecs.components.CollisionComponent
+import no.group15.playmagic.ecs.components.TextureComponent
+import no.group15.playmagic.ecs.components.TransformComponent
+import no.group15.playmagic.events.CollisionEvent
+import java.util.*
+
+
+class CollisionSystem(
+	priority: Int
+) : EntitySystem(
+	priority
+) {
+
+	private lateinit var entities: ImmutableArray<Entity>
+	private val collision = mapperFor<CollisionComponent>()
+	private val transform = mapperFor<TransformComponent>()
+	private val texture = mapperFor<TextureComponent>()
+	private val familiesListenersMap: MutableMap<FamilyPair, Signal<CollisionEvent>> = HashMap()
+
+
+
+	override fun addedToEngine(engine: Engine) {
+		entities = engine.getEntitiesFor(
+			allOf(CollisionComponent::class, TransformComponent::class, TextureComponent::class).get()
+		)
+	}
+
+
+
+	override fun update(deltaTime: Float) {
+
+		for ( (entity1, entity2) in entityPairs() ) {
+
+			if (areColliding(entity1, entity2)) {
+
+				if (isNewCollision(entity1, entity2)) {
+					selectiveDispatch(CollisionEvent(entity1, entity2))
+				}
+
+				// component state update
+				entity1[collision]!!.collidingWith.add(entity2)
+				entity2[collision]!!.collidingWith.add(entity1)
+
+			}
+			else {
+				// component state update
+				entity1[collision]!!.collidingWith.remove(entity2)
+				entity2[collision]!!.collidingWith.remove(entity1)
+			}
+
+		}
+
+	}
+
+
+
+	fun registerListener(family1: Family, family2: Family, listener: Listener<CollisionEvent>) {
+		// TODO: check if FamilyPair equality works
+		familiesListenersMap.getOrPut(FamilyPair(family1, family2), {Signal()}).add(listener)
+	}
+
+
+
+	// -- IMPLEMENTATION --
+
+
+
+	private fun isNewCollision(entity1: Entity, entity2: Entity): Boolean {
+		return ! entity1[collision]!!.collidingWith.contains(entity2)
```

Comment:
When you're returning a 1 liner like this you can use `fun isNewCollision(...) = returnValue` and omit the `return` keyword. It's a bit more concise